src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java

```diff
@@ -174,21 +172,221 @@ public interface UserRepository extends JpaRepository<User, Long> {
+
+                return criteriaBuilder.and(activatedPredicate, deactivatedPredicate);
+            };
+        }
+    }
+
+    /**
+     * Creates the specification to match the users course.
+     *
+     * @return specification used to chain database operations
+     */
+    private Specification<User> getAllUsersMatchingEmptyCourses() {
+        return (root, query, criteriaBuilder) -> criteriaBuilder.isEmpty(root.get(User_.GROUPS));
+    }
+
+    /**
+     * Creates the specification to match the users course.
+     *
+     * @param courseIds a set of courseIds which the users need to match at least one
+     * @return specification used to chain database operations
+     */
+    private Specification<User> getAllUsersMatchingCourses(Set<Long> courseIds) {
+        return (root, query, criteriaBuilder) -> {
+            // Sub query to find all users belonging to a group corresponding to the selected course.
+            Subquery<Long> subQuery = query.subquery(Long.class);
+            Root<User> subUserRoot = subQuery.from(User.class);
+            Root<Course> subCourseRoot = subQuery.from(Course.class);
+            Join<User, String> group = subUserRoot.join(User_.GROUPS, JoinType.LEFT);
+
+            // Select all possible group types
+            Predicate student = criteriaBuilder.in(subCourseRoot.get(Course_.STUDENT_GROUP_NAME)).value(group);
+            Predicate teaching = criteriaBuilder.in(subCourseRoot.get(Course_.TEACHING_ASSISTANT_GROUP_NAME)).value(group);
+            Predicate editor = criteriaBuilder.in(subCourseRoot.get(Course_.EDITOR_GROUP_NAME)).value(group);
+            Predicate instructor = criteriaBuilder.in(subCourseRoot.get(Course_.INSTRUCTOR_GROUP_NAME)).value(group);
+
+            // The course needs to be one of the selected
+            Predicate inCourse = criteriaBuilder.in(subCourseRoot.get(Course_.ID)).value(courseIds);
+
+            subQuery.select(subUserRoot.get(User_.ID))
+                    .where(criteriaBuilder.and(criteriaBuilder.or(student, teaching, editor, instructor), inCourse,
+                            criteriaBuilder.equal(subUserRoot.get(User_.ID), root.get(User_.ID))))
+                    .groupBy(subUserRoot.get(User_.ID)).having(criteriaBuilder.equal(criteriaBuilder.count(group), courseIds.size()));
+
+            Predicate notEmptyAndValidGroups = criteriaBuilder.and(criteriaBuilder.isNotEmpty(root.get(User_.GROUPS)), criteriaBuilder.exists(subQuery));
+
+            return criteriaBuilder.or(notEmptyAndValidGroups);
+        };
+    }
+
+    /**
+     * Creates the specification to match the users course.
+     *
+     * @param courseIds a set of courseIds which the users need to match at least one
+     * @return specification used to chain database operations
+     */
+    private Specification<User> getCourseSpecification(Set<Long> courseIds) {
+        if (courseIds.size() == 1 && courseIds.contains(-1L)) {
```

Comment:
I'm not sure here, but I would suggest to create a constant for "-1L" like `COURSE_ID_FOR_EMPTY_COURSES` that makes the purpose of the -1 clear

Reply:
Good point :D Thank you!